// thu-info-lib/dist/utils/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScheduleError = exports.GitLabApiError = exports.ResponseStatusError = exports.UserInfoError = exports.CabError = exports.IdAuthError = exports.SportsError = exports.LibraryError = exports.EleError = exports.LoseCardError = exports.ClassroomStateError = exports.AssessmentError = exports.ReportError = exports.UrlError = exports.LoginError = exports.LibError = void 0;
class LibError extends Error {
    constructor(message) {
        super(message);
    }
}
exports.LibError = LibError;
class LoginError extends LibError {
    constructor(message) {
        super(message);
    }
}
exports.LoginError = LoginError;
class UrlError extends LibError {
}
exports.UrlError = UrlError;
class ReportError extends LibError {
}
exports.ReportError = ReportError;
class AssessmentError extends LibError {
    constructor(message) {
        super(message);
    }
}
exports.AssessmentError = AssessmentError;
class ClassroomStateError extends LibError {
}
exports.ClassroomStateError = ClassroomStateError;
class LoseCardError extends LibError {
}
exports.LoseCardError = LoseCardError;
class EleError extends LibError {
}
exports.EleError = EleError;
class LibraryError extends LibError {
    constructor(message) {
        super(message);
    }
}
exports.LibraryError = LibraryError;
class SportsError extends LibError {
    constructor(message) {
        super(message);
    }
}
exports.SportsError = SportsError;
class IdAuthError extends LibError {
}
exports.IdAuthError = IdAuthError;
class CabError extends LibError {
}
exports.CabError = CabError;
class UserInfoError extends LibError {
}
exports.UserInfoError = UserInfoError;
class ResponseStatusError extends LibError {
    constructor(message) {
        super(message);
    }
}
exports.ResponseStatusError = ResponseStatusError;
class GitLabApiError extends LibError {
}
exports.GitLabApiError = GitLabApiError;
class ScheduleError extends LibError {
}
exports.ScheduleError = ScheduleError;
