// thu-info-lib/dist/lib/water.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postWaterSubmission = exports.getUserInformationAndStore = exports.getUserInformation = void 0;
const strings_1 = require("../constants/strings");
const network_1 = require("../utils/network");
const getUserInformation = (water, id) => {
    water.id = id;
    return (0, network_1.uFetch)(strings_1.WATER_USER_URL, {
        name: "pw",
        param: water.id,
    }).then(JSON.parse);
};
exports.getUserInformation = getUserInformation;
const getUserInformationAndStore = (water, id) => {
    return (0, exports.getUserInformation)(water, id).then(res => {
        water.address = res.address;
        water.phone = res.phone;
    });
};
exports.getUserInformationAndStore = getUserInformationAndStore;
const postWaterSubmission = (water, num, num1, lid) => (0, network_1.uFetch)(strings_1.WATER_SUB_URL, {
    pw: water.id,
    num: num,
    num1: num1,
    lid: lid,
    address: water.address,
}).then(res => {
    if (!res.includes("成功")) {
        throw new Error("Submitting failed");
    }
});
exports.postWaterSubmission = postWaterSubmission;
