// thu-info-lib/dist/mocks/source/newsHtml.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.newsHtml = exports.url9 = exports.url8 = exports.url7 = exports.url6 = exports.url5 = exports.url4 = exports.url3 = exports.url2 = exports.url1 = exports.url0 = void 0;
const _0_1 = __importDefault(require("./0"));
const _1_1 = __importDefault(require("./1"));
const _2_1 = __importDefault(require("./2"));
const _3_1 = __importDefault(require("./3"));
const _4_1 = __importDefault(require("./4"));
const _5_1 = __importDefault(require("./5"));
const _6_1 = __importDefault(require("./6"));
const _7_1 = __importDefault(require("./7"));
const _8_1 = __importDefault(require("./8"));
const _9_1 = __importDefault(require("./9"));
exports.url0 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421e9fd528569336153301c9aa596522b20735d12f268e561f0/boarddetail_cat.jsp?columnId=0010106&itemSeq=46224";
exports.url1 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421fae0429e207e6b597d469dbf915b243d8ae9128e1cdcffb247/jwcbg/detail_cat.jsp?boardid=57&seq=7632";
exports.url2 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421fae0429e207e6b597d469dbf915b243d8ae9128e1cdcffb247/jwcbg/detail_cat.jsp?boardid=57&seq=7631";
exports.url3 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421fae0429e207e6b597d469dbf915b243d8ae9128e1cdcffb247/jwcbg/detail_cat.jsp?boardid=57&seq=7627";
exports.url4 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421f2fa598421322653770bc7b88b5c2d32530b094045c3bd5cabf3/boarddetail_cat.jsp?columnId=xtw01&itemSeq=124715";
exports.url5 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421f2fa598421322653770bc7b88b5c2d32530b094045c3bd5cabf3/boarddetail_cat.jsp?columnId=xtw01&itemSeq=124712";
exports.url6 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421e2e442d23323615e79009cadd650272001f8dd/rscbg/detail.jsp?boardid=22&seq=5182";
exports.url7 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421e0f852882e3e6e5f301c9aa596522b2043f84ba24ebecaf8/node/279771";
exports.url8 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421f2fa598421322653770bc7b88b5c2d32530b094045c3bd5cabf3/boarddetail_cat.jsp?columnId=dwxcb01&itemSeq=124530";
exports.url9 = "https://webvpn.tsinghua.edu.cn/http/77726476706e69737468656265737421e0f852882e3e6e5f301c9aa596522b2043f84ba24ebecaf8/node/279661";
exports.newsHtml = {};
exports.newsHtml[exports.url0] = _0_1.default;
exports.newsHtml[exports.url1] = _1_1.default;
exports.newsHtml[exports.url2] = _2_1.default;
exports.newsHtml[exports.url3] = _3_1.default;
exports.newsHtml[exports.url4] = _4_1.default;
exports.newsHtml[exports.url5] = _5_1.default;
exports.newsHtml[exports.url6] = _6_1.default;
exports.newsHtml[exports.url7] = _7_1.default;
exports.newsHtml[exports.url8] = _8_1.default;
exports.newsHtml[exports.url9] = _9_1.default;
